dbms/src/Operators/ExchangeReceiverSourceOp.cpp

```diff
@@ -0,0 +1,108 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Operators/ExchangeReceiverSourceOp.h>
+
+namespace DB
+{
+void ExchangeReceiverSourceOp::operateSuffix() noexcept
+{
+    LOG_DEBUG(log, "finish read {} rows from remote", total_rows);
+}
+
+Block ExchangeReceiverSourceOp::popFromBlockQueue()
+{
+    assert(!block_queue.empty());
+    Block block = std::move(block_queue.front());
+    block_queue.pop();
+    return block;
+}
+
+OperatorStatus ExchangeReceiverSourceOp::readImpl(Block & block)
+{
+    if (!block_queue.empty())
+    {
+        block = popFromBlockQueue();
+        return OperatorStatus::HAS_OUTPUT;
+    }
+
+    while (true)
```

Comment:
Because `ExchangeReceiverSourceOp` cannot decode in `awaitImpl`

Reply:
> Because `ExchangeReceiverSourceOp` cannot decode in `awaitImpl`

Yeah. Why not check recv_res's ReceiveStatus in awaitImpl, then decode it in readImpl?